src/sort.rs

```diff
@@ -0,0 +1,174 @@
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use ego_tree::tree;
+    ///
+    /// let mut tree = tree!("1a" => { "2b", "4c", "3d" });
+    /// tree.root_mut().sort_by_key(|a| a.value().split_at(1).0.parse::<i32>().unwrap());
+    /// assert_eq!(
+    ///     vec!["2b", "3d", "4c"],
+    ///     tree.root()
+    ///         .children()
+    ///         .map(|n| *n.value())
+    ///         .collect::<Vec<_>>(),
+    /// );
+    ///
+    /// // Example for sort_by_id.
+    /// tree.root_mut().sort_by_key(|n| n.id());
+    /// assert_eq!(
+    ///     vec![&"2b", &"4c", &"3d"],
+    ///     tree.root()
+    ///         .children()
+    ///         .map(|n| n.value())
+    ///         .collect::<Vec<_>>(),
+    /// );
+    /// ```
+    pub fn sort_by_key<K, F>(&mut self, mut f: F)
+    where
+        F: FnMut(&NodeRef<T>) -> K,
+        K: Ord,
+    {
+        if self.has_children() {
+            let ids = self.sorted_ids(|nodes| {
+                nodes.sort_by_key(|node| f(node));
+            });
+
+            for id in ids {
+                self.append_id(id);
+            }
+        }
+    }
+
+    /// Returns a vector of `NodeId`s of the node's children, sorted according to
+    /// the provided closure.
+    ///
+    /// The closure `f` is applied to a mutable vector of `NodeRef`s representing
+    /// the children of the node. This allows you to define the sorting logic
+    /// within the closure, which will then determine the order of the returned
+    /// `NodeId`s.
+    ///
+    /// # Parameters
+    ///
+    /// - `f`: A closure that takes a mutable reference to a vector of `NodeRef`s.
+    ///         The closure is responsible for sorting the vector in-place.
+    ///
+    /// # Returns
+    ///
+    /// A vector of `NodeId`s representing the children of the node, sorted
+    /// according to the logic defined in the closure `f`.
+    pub fn sorted_ids<F>(&self, mut f: F) -> Vec<NodeId>
+    where
+        F: FnMut(&mut Vec<NodeRef<T>>),
+    {
+        let node_ref = unsafe { self.tree.get_unchecked(self.id) };
+        node_ref.sorted_ids(&mut f)
+    }
+}
+
+impl<'a, T: 'a> NodeRef<'a, T> {
+    /// Returns a vector of `NodeId`s of the node's children, sorted according to
+    /// the provided closure.
+    ///
+    /// The closure `f` is applied to a mutable vector of `NodeRef`s representing
+    /// the children of the node. This allows you to define the sorting logic
+    /// within the closure, which will then determine the order of the returned
+    /// `NodeId`s.
+    ///
+    /// # Parameters
+    ///
+    /// - `f`: A closure that takes a mutable reference to a vector of `NodeRef`s.
+    ///         The closure is responsible for sorting the vector in-place.
+    ///
+    /// # Returns
+    ///
+    /// A vector of `NodeId`s representing the children of the node, sorted
+    /// according to the logic defined in the closure `f`.
+    pub fn sorted_ids<F>(&self, mut f: F) -> Vec<NodeId>

```

Comment:
Pub this method allows people to easily implement their own swap methods.(maybe has a lifetime problem)

Reply:
I'm not sure if it should be named `to_sorted_ids` or `to_sort_ids` in Rust.

Reply:
I am not a fan of exposing this. It is trivial to implement oneself and just exposes API surface that we might not need for a future optimized variant of `sort*`.

Reply:
The reason I initially exposed this interface was because, when I first wrote this code, I couldn't have both `NodeMut` and `NodeRef` at the same time. However, it seems that the borrowing issue might have been resolved after switching to `NodeId`. I will remove it for now.